Adding additional fields to vital types for tracking

aaron-bray opened the pull request:
+ default ctor for bounding box
+ geo_pt to detected_object
+ image_point, tracking_point to object_track_state

linus-sherrill commented on vital/types/object_track_set.h:
```diff
@@ -104,20 +106,7 @@ class VITAL_EXPORT object_track_state : public track_state
   object_track_state( object_track_state&& other ) = default;
 
   /// Clone the track state (polymorphic copy constructor)
-  track_state_sptr clone( clone_type ct = clone_type::DEEP ) const override
```
Some additional comments could help here. What are the option settings and how does that effect the result?

aaron-bray replied:
The options are just deep and shallow, which are pretty standard.

I did move the implementation from the .h to the .cxx, maybe its better to keep in the header?

mwoehlke-kitware replied:
See also #812. Note that the default needs to be `DEEP`, otherwise it is hard to generically and *safely* clone a KWIVER object. As noted, "deep" and "shallow" are generally understood, but I guess if documentation is wanted, it should be on the `enum`. To wit, a shallow clone can potentially be mutated by mutating the original.

I think moving to the `.cxx` is fine, especially as I think it often can't be inlined anyway (often it will get called via virtual dispatch). I think I only had it in the header because it was there prior to #812. (@mleotta might have an opinion?)

linus-sherrill replied:
Generally I prefer the implementation in the .cxx file unless there is some mitigating factors like it is a template.

mwoehlke-kitware commented on vital/types/bounding_box.cxx:
```diff
@@ -1,5 +1,5 @@
 /*ckwg +29
- * Copyright 2016-2017 by Kitware, Inc.
+ * Copyright 2016-2017, 2019 by Kitware, Inc.
```
...but nothing in this file changed??

mwoehlke-kitware commented on vital/types/bounding_box.h:
```diff
@@ -95,6 +95,22 @@ class bounding_box
+  /**
+   * @brief Create default (invalid) box.
+   */
+  bounding_box() : bounding_box( -1, -1, -1, -1 )
```
Are you sure? If I am reading correctly, this creates a box with TL = (-1, -1), W=0, H=0, which is *valid*.

I think you want `: bounding_box{ 0, 0, -1, -1 }`?

aaron-bray replied:
This ctor I am using is setting the upper and lower left/right values (not width and height)
W, H is computed by eigen
Even if I did call what you suggest (using 0 width), I would get an eigen box of all -1's

We kind of talked about what is an invalid box before, and you brought up the point that
 _Can't predictive boxes have corners less than 0? (I think the "safe" way to implement this, assuming Eigen lets you, is to define "invalid" as having a negative width and/or height. I want to say that's how e.g. QRect[F] is implemented...)_

Would any box that has all the same values for each point be invalid?

I am not sure

mwoehlke-kitware replied:
> This ctor I am using is setting the upper and lower left/right values

That is consistent with my reading of the code. Since `is_valid` was previously true for boxes with non-negative width and height, I suggested instead creating a box that would have negative width and height, by means of the bottom and right coordinates being less than top and left, respectively. (It looks like this is the same way Eigen represents "empty".)

Alternatively, AlignedBox actually has an initializing default ctor; you could just use that, and further have `is_valid` just return `!m_bbox.isEmpty()`.

Actually, this made me think of something else; have you considered adding a `reset` method?

aaron-bray replied:
I like this ideal!

mwoehlke-kitware commented on vital/types/detected_object.cxx:
```diff
@@ -38,6 +38,13 @@
+                                  detected_object_type_sptr classifications )
+  : m_confidence( confidence )
+  , m_type( classifications )
+  , m_index( 0 )
```
This (`m_index`) looks like a good candidate for an inline initializer ([[misc.modern.member_init]](../tree/master/doc/code-style.rst))... Also, consider using brace initializers ([[misc.modern.construct]](../tree/master/doc/code-style.rst))?

mwoehlke-kitware commented on vital/types/detected_object.cxx:
```diff
@@ -68,10 +85,27 @@ ::clone() const
   new_obj->m_index = this->m_index;
   new_obj->m_detector_name = this->m_detector_name;
   new_obj->m_descriptor = this->m_descriptor;
+  new_obj->m_geo_point = this->m_geo_point;
```
I *think* this is okay, but please consider: if I clone a `detection`, someone that has the original detection must not be able to modify the clone's geo-point. (Again, I *think* it is okay, I just want to be sure...)

aaron-bray replied:
The geo_point is not a pointer, so it's copying the value to a new object, so the two values are independent of each other

mwoehlke-kitware commented on vital/types/detected_object.h:
```diff
@@ -43,10 +43,11 @@
 #include <vital/types/descriptor.h>
+#include <vital/types/geo_point.h>
 #include <vital/types/image_container.h>
+#include <vital/types/vector.h>
```
[[includes.order]](../tree/master/doc/code-style.rst) ...thanks for fixing! :+1:

mwoehlke-kitware commented on vital/types/detected_object.h:
```diff
@@ -80,6 +81,15 @@ class VITAL_EXPORT detected_object
+   * @param classifications Optional object classification.
+   */
+  detected_object( double confidence = 1.0,
+                   detected_object_type_sptr classifications = detected_object_type_sptr());
```
BTW, `= {}` or `= nullptr`? (Also other new ctor, and maybe change the existing one(s) while you're at it?)

mwoehlke-kitware commented on vital/types/detected_object.h:
```diff
@@ -120,6 +141,27 @@ class VITAL_EXPORT detected_object
+   *
+   * @param gp geo_point for this detection.
+   */
+  void set_geo_point( const kwiver::vital::geo_point& gp );
```
BTW, I see this file was already wrong, but [[types.qualified]](../tree/master/doc/code-style.rst)...

aaron-bray replied:
I corrected this, note that it would be better to link to an rtd page of the code style, for whatever reason, when I click on this link in github this link resolves to : https://github.com/Kitware/kwiver/compare/tree/master/doc/code-style.rst?expand=1

Which is not helpful at all....

mwoehlke-kitware replied:
Unfortunately, RTD isn't going to help much either; at least not directly. There is no easy, non-repetitive way (AFAICT) to generate anchors for non-headings short of writing custom code. If there was, github would equally have anchors and an RTD version wouldn't be needed. Lacking such custom code, an RTD version doesn't have anchors either.

mwoehlke-kitware commented on vital/types/object_track_set.cxx:
```diff
@@ -68,5 +68,28 @@ ::object_track_set( std::vector< track_sptr > const& tracks )
+    copy->set_image_point(this->image_point_);
+    copy->set_track_point(this->track_point_);
+
+    return copy;
```
`return std::move( copy );`

mwoehlke-kitware commented on vital/types/object_track_set.h:
```diff
@@ -48,6 +48,8 @@
 
 #include <vital/range/transform.h>
 
+#include <vital/types/point.h>
```
FYI, `vital/types` s.b. before `vital/range`, and probably before (or maybe with) `vital/*.h`. (`vital/range` is "below" pretty much everything else; they are STL-like headers and have no external dependencies.)

aaron-bray replied:
Ok, I took a shot.
I was just going by the alphabet

mwoehlke-kitware commented on vital/types/object_track_set.h:
```diff
@@ -129,8 +118,35 @@ class VITAL_EXPORT object_track_state : public track_state
+    return detection_;
+  }
 
+  void set_image_point(point_2d const& p )
```
BTW, missing space in `(point` (also below).

aaron-bray commented:
@mwoehlke-kitware I believe this is all good now. Double check the style/structure is good and I addressed your comments well. The last thing I want to do is add these new fields to serialization classes, and hence things would be tested pretty well.

If fields and functions are to your liking, I will start on that and we can call it a day on this

mwoehlke-kitware commented on vital/types/bounding_box.h:
```diff
@@ -95,6 +95,32 @@ class bounding_box
+   */
+  bounding_box()
+  {
+    reset();
```
...pretty sure this is redundant?

aaron-bray replied:
I am having the default ctor call the reset, so they are kept to be the same. 
In case anything new is added, folks should see to default variables in the reset, rather than the ctor (and maybe makine the ctor and reset method out of sync)

mwoehlke-kitware replied:
Hmm, I suppose I can buy that... mostly it bugs me that `setEmpty` is getting called twice.

aaron-bray replied:
Well, I could always just add a comment explaining this and to keep the reset in sync with the ctor

People read comments, right?

mwoehlke-kitware commented on vital/types/bounding_box.h:
```diff
@@ -95,6 +95,32 @@ class bounding_box
+   */
+  void reset()
+  {
+    m_bbox = Eigen::AlignedBox< T, 2 >();
```
I don't know if it matters, but `m_bbox.setEmpty()` *might* be more efficient. (At best, I believe the compiler would optimize this into the equivalent.)

mwoehlke-kitware commented on vital/types/detected_object.cxx:
```diff
@@ -68,10 +82,27 @@ ::clone() const
+// ------------------------------------------------------------------
+void
+detected_object
+::set_geo_point(const kwiver::vital::geo_point& gp)
```
If you end up making other changes... [[types.qualified]](../tree/master/doc/code-style.rst) :wink:

mwoehlke-kitware commented on vital/types/detected_object.h:
```diff
@@ -43,10 +43,11 @@
 #include <vector>
 
 #include <vital/types/detected_object_type.h>
```
If you wind up making other changes, I usually re-sort existing headers also when changing an include list. (`detected_object_type.h` seems to be the only one out of place. If you intend for it to be "file local" as opposed to just "same library", that's okay too, but then it should have a blank line separating it from the others.)

mwoehlke-kitware commented on vital/types/detected_object.h:
```diff
@@ -118,7 +140,28 @@ class VITAL_EXPORT detected_object
+   * @brief Get geo_point from this detection.
+   *
+   * The geo_point for this detection is returned. This point is in
+   * world coordinates. A default constructed (invalid) geo_point
```
Saying that a `geo_point` is in world coordinates is sort of redundant, but eh, if it floats your boat... :slightly_smiling_face:

mwoehlke-kitware commented on vital/types/object_track_set.cxx:
```diff
@@ -68,5 +68,28 @@ ::object_track_set( std::vector< track_sptr > const& tracks )
+    auto copy = std::make_shared< object_track_state >(
+      this->frame(), this->time(), std::move( new_detection ) );
+
+    copy->set_image_point(this->image_point_);
```
Über-pedantic: [[ws.space]](../tree/master/doc/code-style.rst) :wink:. (This is probably the least enforced and most inconsistent, but I do notice that these lines are inconsistent with the surrounding code...)

mwoehlke-kitware commented on vital/types/object_track_set.h:
```diff
@@ -129,8 +117,40 @@ class VITAL_EXPORT object_track_state : public track_state
+    return image_point_;
+  }
+
+  void set_track_point(point_3d const& p )
```
[[ws.space]](../tree/master/doc/code-style.rst) again... more noticeable because you have space on the right but are missing it on the left.

mwoehlke-kitware commented on vital/types/bounding_box.h:
```diff
@@ -95,6 +95,33 @@ class bounding_box
+   */
+  bounding_box()
+  {
+    // Note That any initial state logic here
```
Pedantic: your capitalization (and maybe where you put the line break) is odd. Also, consider: `// NOTE: Any initial ...`; highlighters will tend to mark this differently. (`CAUTION` and `WARNING` work also, though not sure something stronger than `NOTE` is needed.)

mwoehlke-kitware commented on vital/types/detected_object.h:
```diff
@@ -76,9 +79,18 @@ class VITAL_EXPORT detected_object
+   * @param classifications Optional object classification.
+   */
+  detected_object( double confidence = 1.0,
+                   detected_object_type_sptr classifications = nullptr);
```
Nit: missing space inside `)` (also next two ctors)... apologies for not noticing before.
